runtime/stringbuf.c

```diff
@@ -281,6 +281,10 @@ rsRetVal rsCStrAppendStrWithLen(cstr_t *pThis, const uchar* psz, size_t iStrLen)
 	memcpy(pThis->pBuf + pThis->iStrLen, psz, iStrLen);
 	pThis->iStrLen += iStrLen;
 
+	/* reset any cached sz string */
+	free(pThis->pszBuf);
```

Comment:
2016-05-11 9:26 GMT+02:00 Magnus Hyllander <notifications@github.com>:

> In runtime/stringbuf.c
> <https://github.com/rsyslog/rsyslog/pull/1020#discussion_r62802165>:
>
> > @@ -281,6 +281,10 @@ rsRetVal rsCStrAppendStrWithLen(cstr_t *pThis, const uchar* psz, size_t iStrLen)
> >  	memcpy(pThis->pBuf + pThis->iStrLen, psz, iStrLen);
> >  	pThis->iStrLen += iStrLen;
> >
> > +	/* reset any cached sz string */
> > +	free(pThis->pszBuf);
>
> In other words, you want to leave the bug
>

as I said, I need to review this more in-depth. BUT IIRC there is NO BUG in
stringbuf, it is just that the internal API is incorrectly called. Note
that there are many, many more callers of this internal API, and the change
affects many things (e.g. does it create a memleak? How do the many more
NULL writes affect performance especially given the fact that in 99.99% of
the cases it is not needed).

> in the code and work around it elsewhere.
>

IMHO wrong assumption. There is one place that does it wrong. Wouldn't that
be the best place for a fix?


> But it won't work: if all the messages in the file are multiline there
> will be no time you can call persistStrmState without triggering the bug.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/rsyslog/rsyslog/pull/1020/files/342dbca9d61c9b082500167c3f008c0a38720758#r62802165>
>